lib/src/main/java/com/auth0/msg/AbstractMessage.java

```diff
@@ -0,0 +1,180 @@
+package com.auth0.msg;
+
+import com.auth0.jwt.algorithms.Algorithm;
+import org.apache.commons.codec.binary.Hex;
+
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public abstract class AbstractMessage implements Message {
+    private Map<String, Object> claims;
+    private String input;
+    private Error error = null;
+    private boolean verified = false;
+
+    protected AbstractMessage(Map<String, Object> claims) {
+        this.claims = claims;
+    }
+
+    /**
+     * Returns a hashmap representation of the contents of the urlEncoded string
+     * which is passed in as a parameter
+     *
+     * @param urlEncoded the urlEncoded String representation of a message
+     * @return a map of the key value pairs encoded in the string parameter
+     */
+    private static Map<String, Object> claimsFromUrlEncoded(String urlEncoded) throws Exception {
+        //Logic to extract from the string the values
+        Map<String, Object> values = new HashMap<String, Object>();
+        return values;
+    }
+
+    /**
+     * @param input the urlEncoded String representation of a message
+     * @return a Message representation of the UrlEncoded string
+     */
+    public Message fromUrlEncoded(String input) {
+        this.input = input;
+        //This will have logic to parse urlencoded to claims
+        return this;
+    }
+
+    /**
+     * Takes the claims of this instance of the AbstractMessage class and serializes them
+     * to an urlEncoded string
+     *
+     * @return an urlEncoded string
+     */
+    public String toUrlEncoded() {
+        // Serialize the content of this instance (the claims map) into an UrlEncoded string
+        return "";
+    }
+
+    /**
+     * Logic to extract from the string the values
+     *
+     * @param input The JSON String representation of a message
+     * @return a Message representation of the Json
+     */
+    public Message fromJson(String input) {
+        this.input = input;
+        //This will have logic to parse json to claims
+        return this;
+    }
+
+    /**
+     * Takes the claims of this instance of the AbstractMessage class and serializes them
+     * to a json string
+     *
+     * @return a JSON String representation in the form of a hashMap mapping string -> string
+     */
+    public String toJson() {
```

Comment:
All the 'to' methods toJson, toJwt and toUrlEncoded, show have` throws InvalidClaimsException` so that caller will know that it can happen and can handle it.